src/compiler/checker.ts

```diff
@@ -6171,6 +6171,8 @@ namespace ts {
 
         function resolveStructuredTypeMembers(type: StructuredType): ResolvedType {
             if (!(<ResolvedType>type).members) {
+                const earlySymbols = (type.symbol && type.symbol.members) || emptySymbols;

```

Comment:
If `type` is a generic type, its symbol's members will not be the instantiated ones, they'll be the declared ones, right? That would mean this `earlySymbols` table will not have an appropriate type mapper applied to it, if one should be available, which could surface in strange stateful behavior that results in type variable leakage.

Reply:
I think this recursion-guard-set needs to be handled within each sub-`resolve` function to account for this correctly. Actually, they all distill to flavors of `resolveObjectTypeMembers` or `resolveAnonymousTypeMembers`, yes? It should be sufficient to add something like `setStructuredTypeMembers(type, members, callSignatures, constructSignatures, stringIndexInfo, numberIndexInfo);` to above line `5800` and `6068`, right? That way all the non-inherited members are present and resolving. `resolveMappedTypeMembers` already has a similar guard.

Reply:
After much discussion, we got a test of dynamic names to exhibit this failure, or something close. Your suggested fix worked, so that's what I did.